File: 5/5_2.py
# Just a simple abs function
def _abs(n):
    return n if n > 0 else -n
# A weird transformation
def ducci(l: list):
    newlist = []
    for i in range(len(l)-1):
        newlist.append(_abs(l[i+1]-l[i]))
    newlist.append(_abs(l[-1]-l[0]))
    return newlist
# solve
def solve_problem():
    n = int(input())
    l = []
    s: set = {()}
    k = ""
    try:
        for i in range(n):
            k = input()
            l.append(int(k))
    except Exception:
        L = k.split()
        for i in L:
            l.append(int(i))
        
    while True:
        if tuple(l) in s:
            if len(list(filter(lambda x: x != 0, l))) == 0:
                print('ZERO')
                break
            else:
                print('LOOP')
                break
        l = ducci(l)
        s.add(tuple(l))
solve_problem()

File: 5/5_2.cpp
#include <iostream>
#include <vector>
#include <set>
#include <cmath>
#include <algorithm>

using namespace std;

// A simple absolute value function
int _abs(int n) {
    return (n > 0) ? n : -n;
}

// The Ducci sequence transformation
vector<int> ducci(const vector<int>& l) {
    vector<int> newlist;
    int len = l.size();
    for (int i = 0; i < len - 1; ++i) {
        newlist.push_back(_abs(l[i + 1] - l[i]));
    }
    newlist.push_back(_abs(l[len - 1] - l[0])); // last to first
    return newlist;
}

// Solve the problem
void solve_problem() {
    int n;
    cin >> n;
    vector<int> l;
    set<vector<int>> s;
    string k;

    for (int i = 0; i < n; ++i) {
        int temp;
        cin >> temp;
        l.push_back(temp);
    }

    while (true) {
        if (s.find(l) != s.end()) {
            // Check if all elements in the sequence are zero
            if (all_of(l.begin(), l.end(), [](int x){ return x == 0; })) {
                cout << "ZERO" << endl;
            } else {
                cout << "LOOP" << endl;
            }
            break;
        }
        s.insert(l);
        l = ducci(l);
    }
}

int main() {
    solve_problem();
    return 0;
}
